lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -732,66 +733,157 @@ class MetricsEndpointProvider(Object):
+                one relation provided by the charm and using the `prometheus_scrape`
+                interface and, if none or multiple are found, it will raise a
+                ops.model.ModelError exception.
             jobs: an optional list of dictionaries where each
                 dictionary represents the Prometheus scrape
                 configuration for a single job. When not provided, a
                 default scrape configuration is provided for the
-                `/metrics` endpoint pooling using port `80`.
+                `/metrics` endpoint pooling all units of this charm on port `80`.
             alert_rules_path: an optional path for the location of alert rules
-                files.  Defaults to "src/prometheus_alert_rules" at the top level
-                of the charm repository.
+                files.  Defaults to "./prometheus_alert_rules",
+                resolved from the directory hosting the charm entry file.
+                The alert rules are automatically update on charm upgrade.
         """
-        super().__init__(charm, name)
+        if not relation_name:
+            # Check if there is just one relation with the right interface
+            prometheus_scrape_relations = [
+                relation_name
+                for relation_name, relation_meta in charm.meta.provides.items()
+                if relation_meta.interface_name == "prometheus_scrape"
+            ]
+
+            if len(prometheus_scrape_relations) == 1:
+                relation_name = prometheus_scrape_relations[0]
+            elif len(prometheus_scrape_relations) == 0:
+                raise ModelError(
+                    "No provided relation with the 'prometheus_scrape' interface found"
+                )
+            else:
+                raise ModelError(
+                    "Multiple relations with the 'prometheus_scrape' interface found: "
+                    f" {prometheus_scrape_relations}"
+                )
+
+        if not alert_rules_path:
+            import sys
+
+            charm_file = sys.path[0]
+
+            import pathlib
+
+            default_alerts_dir = pathlib.Path(charm_file).joinpath("prometheus_alert_rules")
+
+            if default_alerts_dir.exists() and default_alerts_dir.is_dir:
+                alert_rules_path = str(default_alerts_dir.absolute())
+
+        super().__init__(charm, relation_name)
 
         self._charm = charm
         self._ALERT_RULES_PATH = alert_rules_path
-        self._service_event = service_event
-        self._relation_name = name
+        self._relation_name = relation_name
         # Sanitize job configurations to the supported subset of parameters
         self._jobs = [_sanitize_scrape_configuration(job) for job in jobs]
 
         events = self._charm.on[self._relation_name]
-        self.framework.observe(events.relation_joined, self._set_scrape_job_spec)
+        self.framework.observe(events.relation_created, self._set_scrape_job_spec)
```

Comment:
Setting relation data on created is ok for application data. However quite some time ago I found this to work unreliably. But was a hard to reproduce problem so never filed a bug.  I wonder if this works more consistently now.

Reply:
Agreed. From past experience, "joined" seems safer.